src/profile-logic/flame-graph.js

```diff
@@ -38,229 +40,255 @@ export type FlameGraphTiming = Array<{
 
-    /* From the parent, we can now know the slice allotted for all
-     * its children. */
-    const start = offsets[parent];
-    const end = offsets[parent] + callNodeChildCount[parent] - 1;
-
-    /* Find the place in `array` where this callnode should be
-     * inserted, swapping elements in the array as we go
-     * along. Continue as long as this callnode's function name is
-     * lexically smaller than the function names of the callnodes
-     * already placed in the array. This ensures that all slices have
-     * children in descending order. Any callnode indices equal to 0
-     * means that they are uninitialized, so just breeze through
-     * them. When we stop, when have found the right position to
-     * insert our callnode.
-     *
-     * This effectively is an insertion sort, which is O(n^2), but
-     * since n is typically small (the number of children of a given
-     * callnode), it should be just fine.
-     */
-    let i = start;
-    while (i < end) {
-      if (
-        array[i + 1] !== 0 &&
-        funcName >
-          thread.stringTable.getString(
-            thread.funcTable.name[callNodeTable.func[array[i + 1]]]
-          )
-      ) {
-        // We've found our spot if the next slot in the array is
-        // occupied with a callnode whose function name is less than
-        // ours.
-        break;
+      // Check if the current row contains any other pending call nodes.
+      while (indexInCandidateRow === candidateRow.length) {
+        // There are no more pending nodes in the current row.
+        if (candidateDepth === 0) {
+          // We're completely done.
+          break outer;
+        }
+        // Go up a level and continue the search there.
+        candidateDepth--;
+        candidateRow = callNodesAtDepth[candidateDepth];
+        indexInCandidateRow = pendingRangeStartAtDepth[candidateDepth];
       }
 
-      array[i] = array[i + 1];
-      i++;
+      // We have found a pending call node.
+      candidateNode = candidateRow[indexInCandidateRow];
     }
-    array[i] = callNodeIndex;
-  }
-  offsets[callNodeIndex] = ptr;
 
-  // The children are already sorted, but the roots aren't.
-  // Let's sort the roots in descending order, just like the children.
-  roots.sort((rootA, rootB) => {
-    const [nameA, nameB] = [rootA, rootB].map((callNodeIndex) =>
-      thread.stringTable.getString(
-        thread.funcTable.name[callNodeTable.func[callNodeIndex]]
-      )
-    );
-    if (nameA < nameB) {
-      return 1;
-    }
-    if (nameA === nameB) {
-      return 0;
+    // Now candidateNode is a pending node which has at least one child.
+    // "Finish" candidateNode by incrementing pendingRangeStartAtDepth[candidateDepth].
+    pendingRangeStartAtDepth[candidateDepth] = indexInCandidateRow + 1;
+
+    // Advance to candidateNode's first child.
+    currentCallNode = candidateNode + 1; // "currentCallNode = firstChild[candidateNode];"
```

Comment:
This is the first child because of how the call node is sorted, right? (just checking)

Reply:
correct, I'll tweak the comment